compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp

```diff
@@ -17,6 +19,275 @@ namespace mlir::iree_compiler {
+  // tensor<8x16x8x16xf32> into tensor<8x16x8x16xf32> and then is consumed by
+  // the expanded parallel_insert_slice_op.
+  SmallVector<ReassociationIndices> unFlattenReassociations;
+  for (ReassociationIndices inds : reIndices) {
+    for (int64_t i : inds) {
+      unFlattenReassociations.push_back({i});
+    }
+  }
+  // compute the offsets,sizes,strides in the expanded dimensions.
+  auto computeExpandedAccess = [&](ArrayRef<OpFoldResult> mixedOffsets,
+                                   ShapedType resultType)
+      -> std::tuple<SmallVector<OpFoldResult>, SmallVector<OpFoldResult>,
+                    SmallVector<OpFoldResult>> {
+    SmallVector<OpFoldResult> expandedOffsets;
+    auto expandedOffsetsIter = expandedOffsets.begin();
+
+    for (auto [index, offset] : llvm::enumerate(mixedOffsets)) {
+      // Add zero offsets for the extra dimensions from reIndices.
+      for (size_t i = 1, e = reIndices[index].size(); i < e; ++i) {
+        expandedOffsets.push_back(getAsIndexOpFoldResult(ctx, 0));
+      }
+      // Compute the outer dimension expression.
+      AffineExpr s0, s1;
+      bindSymbols(rewriter.getContext(), s0, s1);
+      AffineExpr outerDimExpr = (s0).floorDiv(s1);
+      // Insert computed offset using affine expression.
+      expandedOffsets.insert(
+          expandedOffsetsIter,
+          affine::makeComposedFoldedAffineApply(
+              rewriter, loc, outerDimExpr,
+              {getValueOrCreateConstantIndexOp(rewriter, loc, offset),
+               rewriter.getIndexAttr(totalInnerSizes[index])}));
+
+      expandedOffsetsIter = expandedOffsets.end();
+    }
+    ArrayRef<int64_t> expandedShape = resultType.getShape();
+    SmallVector<OpFoldResult> expandedSizes;
+    for (int64_t size : expandedShape) {
+      expandedSizes.push_back(getAsIndexOpFoldResult(ctx, size));
+    }
+    SmallVector<OpFoldResult> expandedStrides(resultType.getRank(),
+                                              rewriter.getIndexAttr(1));
+    return {expandedOffsets, expandedSizes, expandedStrides};
+  };
+  for (Operation *user : expandableUsers) {
+    rewriter.setInsertionPointToStart(forallOp.getBody());
+    if (auto extractSliceOp = dyn_cast<tensor::ExtractSliceOp>(user)) {
+      auto expandShapeOp =
+          dyn_cast<tensor::ExpandShapeOp>(*extractSliceOp->getUsers().begin());
+      RankedTensorType resultType = expandShapeOp.getResultType();
+      auto [expandedOffsets, expandedSizes, expandedStrides] =
+          computeExpandedAccess(extractSliceOp.getMixedOffsets(), resultType);
+      rewriter.replaceOpWithNewOp<tensor::ExtractSliceOp>(
+          extractSliceOp, resultType, extractSliceOp.getSource(),
+          expandedOffsets, expandedSizes, expandedStrides);
+      rewriter.setInsertionPoint(expandShapeOp);
+      rewriter.replaceOpWithNewOp<tensor::ExpandShapeOp>(
+          expandShapeOp, resultType, expandShapeOp.getSrc(),
+          unFlattenReassociations);
```

Comment:
I don't think you need this expand_shape. It is just a nop right?

Reply:
I tried replacing users with new extractslice and just erasing it but that was leading to segfaults in future patterns, saw this issue in  three cases
1. this expand
2. the collapse
3. old scf.forall
The IR after the pattern looked good but then other patterns still tried to apply on the erased ops, so i found it was easier to just make them no-ops or leave them with no users and let the canonicalizations do there thing.

Reply:
Actually was just trying this and you are exactly right, if I dont erase then it works, thanks for catching this reduced a lot of unnecessary code.